clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -223,6 +253,69 @@ private void maybeUpdateJoinedSubscription(Set<TopicPartition> assignedPartition
         }
     }
 
+    private void maybeAssignPartitions(final Set<TopicPartition> assignedPartitions) {
+        log.info("Setting newly assigned partitions: {}", Utils.join(assignedPartitions, ", "));
+
+        ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
+        try {
+            listener.onPartitionsAssigned(assignedPartitions);
+        } catch (WakeupException | InterruptException e) {
+            throw e;
+        } catch (Exception e) {
+            log.error("User provided listener {} failed on partition assignment", listener.getClass().getName(), e);
+        }
+    }
+
+    private void maybeLosePartitions(final String rootCause, final Predicate<TopicPartition> predicate) {
+        Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+
+        if (subscriptions.partitionsAutoAssigned() && !ownedPartitions.isEmpty()) {
+            Set<TopicPartition> lostPartitions = metadataSnapshot == null ? ownedPartitions :
+                ownedPartitions.stream().filter(predicate).collect(Collectors.toSet());
+
+            log.info("Lost previously assigned partitions {} due to {}", lostPartitions, rootCause);
+
+            ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
+            try {
+                listener.onPartitionsLost(lostPartitions);
+            } catch (WakeupException | InterruptException e) {
+                throw e;
+            } catch (Exception e) {
+                log.error("User provided listener {} failed on partition being lost", listener.getClass().getName(), e);
+            }
+
+            Set<TopicPartition> leftPartitions = new HashSet<>(ownedPartitions);
+            leftPartitions.removeAll(lostPartitions);
+
+            subscriptions.assignFromSubscribed(leftPartitions);
+        }
+    }
+
+    private void maybeRevokePartitions(final Predicate<TopicPartition> predicate) {
```

Comment:
Could we refactor a bit to let `maybeLosePartitions` and `maybeRevokePartitions` share the same code? They only have one-line difference